src/ray/id.h

```diff
@@ -2,32 +2,57 @@
 #define RAY_ID_H_
 
 #include <inttypes.h>
+#include <limits.h>
 
+#include <chrono>
 #include <cstring>
+#include <mutex>
+#include <random>
 #include <string>
 
 #include "plasma/common.h"
 #include "ray/constants.h"
+#include "ray/util/logging.h"
 #include "ray/util/visibility.h"
 
 namespace ray {
 
-class RAY_EXPORT UniqueID {
+class DriverID;
+class UniqueID;
+
+std::mt19937 RandomlySeededMersenneTwister();
+
+template <typename T>
+class BaseID {
  public:
-  UniqueID();
-  UniqueID(const plasma::UniqueID &from);
-  static UniqueID from_random();
-  static UniqueID from_binary(const std::string &binary);
-  static const UniqueID &nil();
+  BaseID() { std::fill_n(reinterpret_cast<uint8_t *>(this), T::size(), 0xff); }
+
+  static T from_random();
+  static T from_binary(const std::string &binary);
+  static const T &nil();
+  static size_t size() { return sizeof(T); }
+
   size_t hash() const;
   bool is_nil() const;
-  bool operator==(const UniqueID &rhs) const;
-  bool operator!=(const UniqueID &rhs) const;
+  bool operator==(const BaseID &rhs) const;
+  bool operator!=(const BaseID &rhs) const;
   const uint8_t *data() const;
-  static size_t size();
   std::string binary() const;
   std::string hex() const;
-  plasma::UniqueID to_plasma_id() const;
+
+ protected:
+  BaseID(const std::string &binary) {
+    std::memcpy(reinterpret_cast<uint8_t *>(this), binary.data(), T::size());
+  }
+};
+
+#pragma pack(push, 1)
```

Comment:
What is this pragma for? It should probably be documented.

Reply:
This means push current compiler alignment strategy to stack and change current compiler alignment to 1 byte. The default compiler alignment is 8 bytes for x64 machines. `#pragma pack(pop)` will restore the previous compiler alignment strategy. 

If I do not use this, when I set TaskID to 12 bytes and put it to ObjectID, there will be always 4 byte empty data to make it align with 16 bytes. 

I will add a comment to it.